### PR TITLE
5015 live breakeven

### PR DESCRIPTION
**Pull request by X0nic**

### CODE REVIEW

**Comment by X0nic** [comment]
Hey @hooverlunch and @linesarefuzzy , can you take a look at this? I have all the basics working, I just need smash some bugs, and clean it up. Just curious if you think this is a reasonable approach.

**Comment by hooverlunch** [comment]
I will leave this to you and Fuzzy for now since he paired on it, right?
Happy to consult on something specific.

File: app/assets/javascripts/backbone/views/editable_table_view.js.coffee [hunk 1]
```diff
@@ -2,22 +2,46 @@ class MS.Views.EditableTableView extends Backbone.View
...
+
+    console.log({totalFixedCosts: @totalFixedCosts()})
+
+    products = @$("tr[data-group='product']:not('.hidden')").map (index, productRow) =>
+      new MS.Views.BreakevenProductView(el: productRow)
+
+    @$("tr[data-group='product-total']").map (index, totalRow) =>
+      new MS.Views.BreakevenProductTotalView(el: totalRow, products: products)
+
+  totalFixedCosts: =>
```
Comment (X0nic) [inline app/assets/javascripts/backbone/views/editable_table_view.js.coffee]: @linesarefuzzy @hooverlunch Any suggestions on how to get the `totalFixedCosts`? This method isn't going to work because it adds all fixed costs for all break even reports. I can't think of any good ways to do that without subclassing `editable_table` for each of the different tables, and then wiring the two up with events.

**Comment by hooverlunch** [comment]
Again, I'm leaving this to you and @linesarefuzzy. Please understand it
would take time for me to get my head into this code and offer you a
recommendation and that's not a good use of my time if you two can figure
it out on your own. If you as a pair need advice then I'm happy to weigh in.

File: app/assets/javascripts/backbone/views/editable_breakeven_table_view.js.coffee [hunk 2]
```diff
@@ -0,0 +1,27 @@
+class MS.Views.EditableBreakevenTableView extends Backbone.View
```
Comment (X0nic) [inline app/assets/javascripts/backbone/views/editable_breakeven_table_view.js.coffee]: I was unable to get `extends MS.Views.EditableTableView` working here. It is mostly duplicated from `EditableTableView`

File: app/views/admin/loans/questionnaires/_breakeven_product_row.html.slim [hunk 3]
```diff
@@ -3,12 +3,18 @@
...
   td: textarea name="#{prefix}[description]" = product[:description]
   td: textarea name="#{prefix}[unit]" placeholder=I18n.t('loan.breakeven.placeholders.unit') = product[:unit]
-  td: input name="#{prefix}[price]" type="number" value=product[:price]
-  td: input name="#{prefix}[cost]" type="number" value=product[:cost]
-  td: input name="#{prefix}[quantity]" type="number" value=product[:quantity]
+  td: input.price name="#{prefix}[price]" type="number" value=product[:price]
+  td: input.cost name="#{prefix}[cost]" type="number" value=product[:cost]
+  td: input.percentage_of_sales name="#{prefix}[percentage_of_sales]" type="number" value=product[:percentage_of_sales]
+  td
+    input.quantity type="number" value=product[:quantity] disabled=true
```
Comment (X0nic) [inline app/views/admin/loans/questionnaires/_breakeven_product_row.html.slim]: I fought with styling a div to look the same for awhile and wasn't able to do it. So I kept them as inputs, but disabled them.

**Comment by X0nic** [comment]
This took a ton of time because of all of the various interactions with the Dom. Really starting to appreciate what React brings to the table for this type of work.

**Comment by X0nic** [comment]
Hey @hooverlunch this is ready for review. Not sure if you or @linesarefuzzy are doing the review. Feel free to re-assign it if that is the case.

**Comment by linesarefuzzy** [comment]
@hooverlunch FYI I paired with Nathan on the issue he was having and we worked it out.

File: app/assets/javascripts/backbone/views/breakeven_product_total_view.coffee [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  #   'click tr [data-action="delete"]': 'removeRow'
+  #   'click .actions [data-action="add"]': 'addRow'
+
+  initialize: (options) ->
+    @products = options.products
+
+    @totalFixedCosts = parseFloat(@$('.total_fixed_costs').val())
+
+    _.each @products, (product) =>
+      product.on "product:changed", () =>
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/breakeven_product_total_view.coffee]: No need for `()`.

Also, why is this not done using the View's event hash? There should be no need to iterate over all the products like this. That is the point of event bubbling.

File: app/assets/javascripts/backbone/views/breakeven_product_total_view.coffee [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        @calculateTotals()
+
+    # Defer loading of totals until the dom is ready
+    $ =>
+      @calculateTotals()
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/breakeven_product_total_view.coffee]: This whole view should not be loaded until the DOM is ready, right? Why is this needed?

File: app/assets/javascripts/backbone/views/breakeven_product_total_view.coffee [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    _.reduce(@products, (acc, product) =>
+      acc += product.percentageOfSales() if product.isValid()
+
+      return acc
+    , 0)
+
+  Q: ->
+    @totalFixedCosts / @totalPs()
+
+  notifyProducts: () ->
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/breakeven_product_total_view.coffee]: No need for `()`.

File: app/assets/javascripts/backbone/views/breakeven_product_total_view.coffee [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    totalQuantity = @totalQuantity()
+    totalRevenue = @totalRevenue()
+    totalPercentageOfSales = @totalPercentageOfSales()
+    totalProductCost = @totalProductCost()
+    Q = @Q()
+
+    totals = {totalPrice, totalCost, totalQuantity, totalRevenue, totalPercentageOfSales, totalProductCost, @totalFixedCosts, Q}
+    @writeTotalsToDom(totals)
+
+  totalPrice: ->
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/breakeven_product_total_view.coffee]: In future, methods like this that are almost total copies of each other should be abstracted. I bet copy and paste was used to write these methods. That is a big warning sign! The only thing that differs across each is the method called on `product`. I am not happy about this but there isn't time to refactor now unless you think you can do it quickly.

File: app/assets/javascripts/backbone/views/breakeven_product_total_view.coffee [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    _.each @products, (product) =>
+      product.totalsUpdated({@totalFixedCosts, Q: @Q()})
+
+  updated: (totals) ->
+    @totalFixedCosts = totals.totalFixedCosts
+
+    @calculateTotals()
+
+  writeTotalsToDom: (totals) ->
+    @$('.percentage_of_sales').val("#{totals.totalPercentageOfSales * 100} %")
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/breakeven_product_total_view.coffee]: Should not be a space before %, right?

File: app/assets/javascripts/backbone/views/breakeven_product_view.coffee [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+    @updateDom()
+
+  updateDom: ->
+    @writeToDom('revenue', @revenue())
+    @writeToDom('quantity', @quantity())
+    @writeToDom('quantity_display_value', @quantity())
+    @writeToDom('total_cost', @totalCost())
+
+  isValid: =>
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/breakeven_product_view.coffee]: You shouldn't need `=>` on these methods. Note e.g. that `updateDom` doesn't have it but `@` stilll works.

File: app/assets/javascripts/backbone/views/breakeven_product_view.coffee [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    value = @$(".#{fieldName}").val()
+    parseFloat(value)
+
+  writeToDom: (fieldName, value) ->
+    valueForDom = @getValueForDom(value)
+    @$(".#{fieldName}").val(valueForDom).text(valueForDom)
+
+  getValueForDom: (value) =>
+    return value.toFixed() if @isValid()
+    ""
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/breakeven_product_view.coffee]: `if @isValid() then value.toFixed() else ""` would be cleaner.

File: app/assets/javascripts/backbone/views/breakeven_product_view.coffee [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+  totalsUpdated: (totals) ->
+    @total_fixed_costs = totals.totalFixedCosts
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/breakeven_product_view.coffee]: Please stick to coding conventions! camelCase variables!

File: app/assets/javascripts/backbone/views/breakeven_product_view.coffee [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    @Q = totals.Q
+
+    @updateDom()
+
+  changed: () ->
+    @_price = @readFromDom('price')
+    @_cost = @readFromDom('cost')
+    @_percentageOfSales = @readPercentageOfSalesFromDom()
+    @updateDom()
+    @trigger("product:changed", @)
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/breakeven_product_view.coffee]: Single quotes for non-interpolated strings please.

File: app/assets/javascripts/backbone/views/breakeven_view.coffee [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+  initialize: (e) ->
+    @$(".editable-breakeven-table").each (index, table) =>
+      new MS.Views.EditableBreakevenTableView(el: table)
+
+    @products = @$("tr[data-group='product']:not('.hidden')").map (index, productRow) =>
+      new MS.Views.BreakevenProductView(el: productRow)
+
+    totals = @$("tr[data-group='product-total']").map (index, totalRow) =>
+      new MS.Views.BreakevenProductTotalView(el: totalRow, products: @products)
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/breakeven_view.coffee]: This is looking very nice.

File: app/assets/javascripts/backbone/views/breakeven_view.coffee [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+    # Should be 1 and only 1 total row
+    @total = totals[0]
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/breakeven_view.coffee]: This seems an odd way to do things if there is only 1. Why not use jQuery's `.first()` above? `@total = new MS.Views.BreakevenProductTotalView(el: @$("tr[data-group='product-total']").first(), products: @products)`

File: app/assets/javascripts/backbone/views/breakeven_view.coffee [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+
+  totalFixedCosts: =>
+    _.reduce(@$('.editable-table[data-table="fixed_costs"] input.amount'), (acc, amount) =>
+      value = parseFloat($(amount).val())
+      acc += value unless isNaN(value)
+
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/breakeven_view.coffee]: Remove blank line please.

File: app/assets/javascripts/backbone/views/breakeven_view.coffee [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+    totalFixedCosts = @totalFixedCosts()
+    Q = @totalFixedCosts() / @total.totalPs()
+
+    costs = {totalFixedCosts: @totalFixedCosts(), Q: Q}
+    _.each @products, (product) =>
+      product.totalsUpdated(costs)
+
+    @total.updated(costs)
+
+
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/breakeven_view.coffee]: Remove extra blank line please.

File: app/assets/javascripts/backbone/views/breakeven_view.coffee [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+  totalFixedCosts: =>
+    _.reduce(@$('.editable-table[data-table="fixed_costs"] input.amount'), (acc, amount) =>
+      value = parseFloat($(amount).val())
+      acc += value unless isNaN(value)
+
+      return acc
+    , 0)
+
+  totalFixedCostsChanged: =>
+    totalFixedCosts = @totalFixedCosts()
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/breakeven_view.coffee]: This variable is not used?

File: app/assets/javascripts/backbone/views/breakeven_view.coffee [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+
+  totalFixedCosts: =>
+    _.reduce(@$('.editable-table[data-table="fixed_costs"] input.amount'), (acc, amount) =>
+      value = parseFloat($(amount).val())
+      acc += value unless isNaN(value)
+
+      return acc
+    , 0)
+
+  totalFixedCostsChanged: =>
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/breakeven_view.coffee]: No need for `=>`s

File: app/assets/javascripts/backbone/views/breakeven_view.coffee [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+    totalFixedCosts = @totalFixedCosts()
+    Q = @totalFixedCosts() / @total.totalPs()
+
+    costs = {totalFixedCosts: @totalFixedCosts(), Q: Q}
+    _.each @products, (product) =>
+      product.totalsUpdated(costs)
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/breakeven_view.coffee]: Why not use built in JS `forEach`? Also this can/should be a one-liner IMO.

File: app/assets/javascripts/backbone/views/editable_breakeven_table_view.js.coffee [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+  # The view is called from the loan questionnaires view.
+
+  events:
+    'click tr [data-action="delete"]': 'removeRow'
+    'click .actions [data-action="add"]': 'addRow'
+
+  initialize: (e) ->
+    @$el.find('tbody').sortable({
+      handle: "[data-action='move']"
+    })
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/editable_breakeven_table_view.js.coffee]: Don't need curly braces here.

File: app/assets/javascripts/backbone/views/editable_breakeven_table_view.js.coffee [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+
+  addRow: (e) ->
+    e.preventDefault()
+    $button = @$(e.currentTarget)
+    $table = $button.closest('table')
+    $new_row = $table.find('tr.hidden').clone()
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/editable_breakeven_table_view.js.coffee]: camelCase. Please also refactor the other editable table class to use camel case.

File: app/assets/javascripts/backbone/views/breakeven_product_total_view.coffee [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+
+      return acc
+    , 0)
+
+  Q: ->
+    @totalFixedCosts / @totalPs()
+
+  notifyProducts: () ->
+    _.each @products, (product) =>
+      product.totalsUpdated({@totalFixedCosts, Q: @Q()})
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/breakeven_product_total_view.coffee]: Use built-in `forEach`?

File: app/assets/javascripts/backbone/views/editable_breakeven_table_view.js.coffee [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+    $new_row.removeClass('hidden')
+    $table.find("tr[data-group='product-total']").before($new_row)
+    new MS.Views.BreakevenProductView(el: $new_row)
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/editable_breakeven_table_view.js.coffee]: Doesn't this need to get added to the total view? Else won't it not be included in the totals?

File: app/assets/javascripts/backbone/views/editable_breakeven_table_view.js.coffee [hunk 24]
```diff
@@ -0,0 +1,27 @@
...
+
+  removeRow: (e) ->
+    e.preventDefault()
+    $row = @$(e.currentTarget).closest('tr')
+    $row.remove()
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/editable_breakeven_table_view.js.coffee]: Combine this with previous line please.

File: app/assets/javascripts/backbone/views/editable_table_view.js.coffee [hunk 25]
```diff
@@ -2,22 +2,25 @@ class MS.Views.EditableTableView extends Backbone.View
...
-  el: "table.editable-table"
 
-  editableTableInit: (e) ->
-    $('.editable-table tbody').sortable({
+  events:
+    'click tr [data-action="delete"]': 'removeRow'
+    'click .actions [data-action="add"]': 'addRow'
+
+  initialize: (e) ->
+    @$el.find('tbody').sortable({
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/editable_table_view.js.coffee]: Thanks for this cleanup.

File: app/assets/stylesheets/admin/loans/_tables.scss [hunk 26]
```diff
@@ -19,7 +19,7 @@ table.breakeven-table-rampup {
   }
 }
 
-.editable-table {
+.editable-table, .editable-breakeven-table {
```
Comment (hooverlunch) [inline app/assets/stylesheets/admin/loans/_tables.scss]: Can we at least just use the one class here? I don't see a need for a separate one.

File: app/views/admin/loans/questionnaires/_breakeven_table_question.html.slim [hunk 27]
```diff
@@ -1,28 +1,33 @@
-.table-responsive.editable-tables
+.table-responsive.editable-tables.breakeven-tables
 
   .table-container
     h4.title = t("loan.breakeven.products", count: 2)
 
-    table.table.table-condensed.table-striped.table-bordered.editable-table data-table="products"
+    table.table.table-condensed.table-striped.table-bordered.editable-breakeven-table data-table="products"
```
Comment (hooverlunch) [inline app/views/admin/loans/questionnaires/_breakeven_table_question.html.slim]: Don't need a separate class here. Just use the fact that it's inside `.breakeven-tables` to select it.

File: app/assets/javascripts/backbone/views/breakeven_product_total_view.coffee [hunk 28]
```diff
@@ -0,0 +1,104 @@
...
+
+  updated: (totals) ->
+    @totalFixedCosts = totals.totalFixedCosts
+
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/breakeven_product_total_view.coffee]: Remove blank line.

File: app/assets/javascripts/backbone/views/breakeven_view.coffee [hunk 29]
```diff
@@ -0,0 +1,39 @@
+class MS.Views.BreakevenView extends Backbone.View
+
+  # The view is called from the loan questionnare view.
+
+  events:
+    'change .editable-table[data-table="fixed_costs"] input.amount': 'totalFixedCostsChanged'
+
+  initialize: (e) ->
+    @$(".editable-breakeven-table").each (index, table) =>
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/breakeven_view.coffee]: Single quotes. Please fix throughout.

File: app/assets/stylesheets/admin/loans/_tables.scss (hunk 26 above)
Comment (X0nic) [inline app/assets/stylesheets/admin/loans/_tables.scss]: The main reason for the separate class is for the breakeven table to attach the backbone view to. I opted for this, instead of using a not() selector in loan_questionnaires_view.

File: app/assets/javascripts/backbone/views/breakeven_product_total_view.coffee (hunk 5 above)
Comment (X0nic) [inline app/assets/javascripts/backbone/views/breakeven_product_total_view.coffee]: I am all for a better solution to this, but if I don't do this. Sometimes the calculations do not wire up properly.

**Comment by hooverlunch** [comment]
Ok. Perhaps reflect this in the comment. I'm not sure if it's the DOM that
isn't ready. Might be something one of the other views is doing.

File: app/assets/javascripts/backbone/views/breakeven_product_total_view.coffee (hunk 22 above)
Comment (X0nic) [inline app/assets/javascripts/backbone/views/breakeven_product_total_view.coffee]: I think because of the way the array is passed in. I could cast it to an array. `Array(@products).forEach` Would this be preferrable over underscore?

File: app/views/admin/loans/questionnaires/_breakeven_table_question.html.slim (hunk 27 above)
Comment (X0nic) [inline app/views/admin/loans/questionnaires/_breakeven_table_question.html.slim]: That part makes sense, but the problem is that I don't want the EditableView to also attach to this table. With that solution I would end up with two event handlers for `addRow`.

File: app/assets/javascripts/backbone/views/breakeven_product_total_view.coffee (hunk 4 above)
Comment (X0nic) [inline app/assets/javascripts/backbone/views/breakeven_product_total_view.coffee]: Thanks for reminding me. As for the even hash, I was unaware you can use it to attach events to non-dom objects. What is the syntax to do that? Google was unhelpful for me.

I changed it to: 
```coffeescript
@products.on 'product:changed', =>
  @calculateTotals
```

File: app/assets/javascripts/backbone/views/breakeven_product_total_view.coffee (hunk 5 above)
Comment (X0nic) [inline app/assets/javascripts/backbone/views/breakeven_product_total_view.coffee]: I think I can remove this now. This problem might have gone away once I finished wiring everything up properly.

File: app/assets/javascripts/backbone/views/breakeven_product_view.coffee (hunk 9 above)
Comment (X0nic) [inline app/assets/javascripts/backbone/views/breakeven_product_view.coffee]: Sorry, been writing a bunch of ES6 lately. Is it preferable to use `->` when able?

File: app/assets/javascripts/backbone/views/editable_breakeven_table_view.js.coffee (hunk 23 above)
Comment (X0nic) [inline app/assets/javascripts/backbone/views/editable_breakeven_table_view.js.coffee]: Good catch, sounds like a bug. Might have to add a new event to pass this back to the totals.

File: app/views/admin/loans/questionnaires/_breakeven_table_question.html.slim (hunk 27 above)
Comment (hooverlunch) [inline app/views/admin/loans/questionnaires/_breakeven_table_question.html.slim]: Ok... seems kind of like a smell but fine for now.

File: app/assets/javascripts/backbone/views/breakeven_product_total_view.coffee (hunk 4 above)
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/breakeven_product_total_view.coffee]: I guess I've never tried this particular thing, but you should be able to trigger the event on the DOM object associated with the Product View. All Backbone views have DOM elements. So if `myview.trigger('foo')` doesn't trigger an event on the DOM element (not sure, it might or might not), then `myview.el.trigger('foo')` definitely should. Then you can make use of the DOM's built-in event bubbling. In your event hash, assuming the listening view is attached to an element that wrapps the Product elements, you should just need to do e.g. `'product:changed': 'myfunc'`.
